### ISSUE TITLE
cq:close() from within cq causes segfault

### ISSUE THREAD

**Issue by daurnimator** [wahern/cqueues#175]
```lua
local cqueues = require "cqueues"
local cq = cqueues.new()
cq:wrap(function()
	cq:close()
end)
assert(cq:loop())
```

**Comment by wahern** [comment]
Now I remember why I hadn't implemented cqueue:close, and why I had a nagging feeling that simply reusing cqueue__gc as the close method was problematic.

Commit 96bc2f2f contains an interim fix, but it's not nearly sufficient. A close could be called from code run when we invoke pollfd and similar methods on objects. And theoretically it could even be called from the gc handler of any random object collected during some other Lua state operations. cstack_isrunning isn't sufficient to tell us when close is safe to call.

I'm keeping this bug open, but tagging a release to appease the debian gods.

**Comment by daurnimator** [comment]
> Now I remember why I hadn't implemented cqueue:close, and why I had a nagging feeling that simply reusing cqueue__gc as the close method was problematic.

Yeah sorry about this :( I didn't even think about this code path until I wrote some code and accidentally did it in a test case.
